### PR TITLE
Point at type in `let` assignment on type errors

### LINKED ISSUE #61067: Point at cause of expectation for type mismatch on `let` assignments with explicit type
Given

```
fn foo() -> i32 {
    let s: i32 = 3usize;
    s
}
```

we present

```
error[E0308]: mismatched types
 --> src/lib.rs:2:18
  |
2 |     let s: i32 = 3usize;
  |                  ^^^^^^ expected i32, found usize
help: change the type of the numeric literal from `usize` to `i32`
  |
2 |     let s: i32 = 3i32;
  |                  ^^^^
```

We would ideally also point at the assignment's type in the same way we do for type mismatches against a return type:

```
error[E0308]: mismatched types
 --> src/lib.rs:2:18
  |
2 |     let s: i32 = 3usize;
  |            ---   ^^^^^^ expected i32, found usize
  |            |
  |            expected due to this
help: change the type of the numeric literal from `usize` to `i32`
  |
2 |     let s: i32 = 3i32;
  |                  ^^^^
```

### PR DESCRIPTION
**Pull request by estebank**
Fix #61067.

### CODE REVIEW

File: src/test/ui/terminal-width/non-whitespace-trimming-unicode.stderr [hunk 1]
```diff
@@ -1,8 +1,10 @@
 error[E0308]: mismatched types
   --> $DIR/non-whitespace-trimming-unicode.rs:4:415
    |
-LL | ...♯♰♱♲♳♴♵♶♷♸♹♺♻♼♽♾♿⚀⚁⚂⚃⚄⚅⚆⚈⚉4"; let _: () = 42;  let _: &str = "🦀☀☁☂☃☄★☆☇☈☉☊☋☌☍☎☏☐☑☒☓  ☖☗☘☙☚☛☜☝☞☟☠☡☢☣☤☥☦☧☨☩☪☫☬☭☮☯☰☱☲☳☴☵☶☷☸☹☺☻☼☽☾☿♀♁♂♃♄♅♆...
-   |                                               ^^ expected (), found integer
+LL | ...♬♭♮♯♰♱♲♳♴♵♶♷♸♹♺♻♼♽♾♿⚀⚁⚂⚃⚄⚅⚆⚈⚉4"; let _: () = 42;  let _: &str = "🦀☀☁☂☃☄★☆☇☈☉☊☋☌☍☎☏☐☑☒☓  ☖☗☘☙☚☛☜☝☞☟☠☡☢☣☤☥☦☧☨☩☪☫☬☭☮☯☰☱☲☳☴☵☶☷☸☹☺☻☼☽☾☿♀♁♂♃...
+   |                                             --   ^^ expected (), found integer
+   |                                                      |
+   |                                             expected due to this
```
Comment (estebank) [inline src/test/ui/terminal-width/non-whitespace-trimming-unicode.stderr]: This uncovered an actual bug! Exciting!

Comment (Centril) [reply src/test/ui/terminal-width/non-whitespace-trimming-unicode.stderr]: File a FIXME issue?

File: src/librustc_typeck/check/demand.rs [hunk 2]
```diff
@@ -120,6 +120,18 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+        if let Some(hir::Node::Local(hir::Local {
+            ty: Some(ty),
+            init: Some(init),
+            ..
+        })) = self.tcx.hir().find(parent) {
+            if init.hir_id == expr.hir_id {
+                // Point at `let` assignment type.
+                err.span_label(ty.span, "expected due to this");
+            }
+        }
```
Comment (Centril) [inline src/librustc_typeck/check/demand.rs]: Would prefer to have this in its own method like with the `self.suggest_foobar` methods below; e.g. `self.expected_due_to_type_on_let(expr)`.

**Comment by Centril** [comment]
r? @Centril 

r=me with refactoring, issue filed, and rustdoc test adjusted

File: src/test/ui/terminal-width/non-whitespace-trimming-unicode.stderr (hunk 1 above)
Comment (estebank) [reply src/test/ui/terminal-width/non-whitespace-trimming-unicode.stderr]: https://github.com/rust-lang/rust/issues/66552

**Comment by Centril** [comment]
Failed in https://github.com/rust-lang/rust/pull/66601#issuecomment-557050785, @bors r-

Use `--bless --compare-mode=nll` to fix the issue.

**Comment by Centril** [comment]
@bors r+

@estebank when you force-push I don't get a notification ;)

**Pull request merged**